src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -315,7 +308,9 @@ public struct AsyncTaskMethodBuilder<TResult>
     {
 #if !PROJECTN
         /// <summary>A cached task for default(TResult).</summary>
-        internal readonly static Task<TResult> s_defaultResultTask = AsyncTaskCache.CreateCacheableTask(default(TResult)!); // TODO-NULLABLE-GENERIC
+        internal readonly static Task<TResult> s_defaultResultTask = (null != (object)default(TResult)! && typeof(TResult) == typeof(VoidTaskResult)) ?
```

Comment:
You are saving one singleton at the cost of adding more code. Is it a net win?

Reply:
Likely not, as it would be evaluated for every type and it isn't used for the one its specialized for.